x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/TransportDeleteEnrichPolicyAction.java

```diff
@@ -54,6 +66,22 @@ protected AcknowledgedResponse read(StreamInput in) throws IOException {
     @Override
     protected void masterOperation(Task task, DeleteEnrichPolicyAction.Request request, ClusterState state,
                                    ActionListener<AcknowledgedResponse> listener) throws Exception {
+        List<PipelineConfiguration> pipelines = IngestService.getPipelines(state);
+        for (PipelineConfiguration pc: pipelines) {
+            Pipeline pipeline = ingestService.getPipeline(pc.getId());
+
+            for (Processor processor: pipeline.getProcessors()) {
```

Comment:
This might be a little bit of a conundrum in that the processors returned here could be complex processors that contain processors within them. For instance: The ForEachProcessor could contain an ExactMatchProcessor that is called on each value in field.

A quick fix to get more coverage would be to use 
```suggestion
            for (Processor processor: pipeline.flattenAllProcessors()) {
```
But that only flattens processors from `CompoundProcessors` which things like `ForEachProcessor` are not subclassed... Maybe we need to introduce an interface that allows for flattening all sub-processors when a processor contains any?

Reply:
The `ForEachProcessor` and `ConditionalProcessor` are wrapping processor, meaning that they wrap other processor implementations. In both cases the `.getProcessor()` should return the processor that is wrapped. (I think those are the only 2 wrappers) ... so flattenAllProcessors should work for those.

However, if a wrapper processor wraps another wrapper  e.g. `if` on a `ForEachProcessor` will result in `ConditionalProcessor` wrapping a `ForEachProcessor` which would miss the real processor. I think if add 
```
if (processor instanceof ConditionalProcessor) {
                    processor = ((ConditionalProcessor) processor)
```
in the loop it handle that edge case. 

Other then that edge case, i think @jbaiera suggestion to use `flattenAllProcessors()` will work.  Also, it may make sense to add ^^ to `flattenAllProcessors()` since a conditional processor wrapping the real processor is an implementation detail.

Reply:
ug... I just took a look deeper look and `getProcessor()` is not called by the `flattenAllProcessors()`  ... so your loop would need to also take into account `ForEachProcessor` 
```
if (processor instanceof ConditionalProcessor) {
                    processor = ((ConditionalProcessor) processor).getProcessor()
if (processor instanceof ForEachProcessor) {
                    processor = ((ForEachProcessor) processor).getProcessor()
```
We may want to consider adding in a interface like `WrappedProcessor` that exposes `getProcessor()` and possibly handle the unwrapping inside  `flattenAllProcessors(boolean unWrap)`

Reply:
all of this conversation was handled in 4ec68ec2e654397a74d9935af1802db07f3f8ba2